join.go

```diff
@@ -28,23 +28,13 @@ var ErrSymlinkLoop = errors.Wrap(syscall.ELOOP, "secure join")
 // accessed does not exist (or path components don't exist). This is
 // effectively a more broad version of os.IsNotExist.
 func IsNotExist(err error) bool {
-	// If it's a bone-fide ENOENT just bail.
-	if os.IsNotExist(errors.Cause(err)) {
+	// If it's a bone-fide ENOENT just bail, also check if it's not actually
+	// an ENOTDIR, which in some cases is a more convoluted case of ENOENT
+	// (usually involving weird paths).
+	if errors.Is(err, os.ErrNotExist) || errors.Is(err, syscall.ENOTDIR) || errors.Is(err, syscall.ENOENT) {
```

Comment:
I would do
```go
return os.IsNotExist(err) || errors.Is(err, syscall.ENOTDIR) || errors.Is(err, syscall.ENOENT)
```

(mostly because `os.IsNotExist` could do more than `errors.Is(err, os.ErrNotExist)`.

Reply:
I take it back; as @thaJeztah pointed out elsewhere, the official docs call for using `errors.Is(err, os.ErrNotExist)`.